lib/plugins/paragon-webpack-plugin/ParagonWebpackPlugin.js

```diff
@@ -0,0 +1,230 @@
+      }
+      themeVariantCssAssets[themeVariant] = {
+        fileName: foundThemeVariantAsset.name,
+        default: value.default,
+        dark: value.dark,
+      };
+    });
+
+    if (!coreCssAsset || !Object.keys(themeVariantCssAssets).length === 0) {
+      return {
+        coreCssAsset: undefined,
+        themeVariantCssAssets: {},
+      };
+    }
+
+    return {
+      coreCssAsset: {
+        fileName: coreCssAsset?.name,
+      },
+      themeVariantCssAssets,
+    };
+  }
+
+  findScriptInsertionPoint({ document, originalSource }) {
+    const bodyElement = this.getDescendantByTag(document, 'body');
+    if (!bodyElement) {
+      throw new Error('Missing body element in index.html');
+    }
+
+    // determine script insertion point
+    if (bodyElement.sourceCodeLocation?.endTag) {
+      return bodyElement.sourceCodeLocation.endTag.startOffset;
+    }
+
+    // less accurate fallback
+    return originalSource.indexOf('</body>');
+  }
+
+  minifyScript(script) {
+    return script
+      .replace(/>[\r\n ]+</g, '><')
+      .replace(/(<.*?>)|\s+/g, (m, $1) => {
+        if ($1) { return $1; }
+        return ' ';
+      })
+      .trim();
+  }
+
+  insertParagonScriptIntoDocument({
+    originalSource,
+    scriptContents,
+  }) {
+    // parse file as html document
+    const document = parse5.parse(originalSource, {
+      sourceCodeLocationInfo: true,
+    });
+
+    // find the body element
+    const scriptInsertionPoint = this.findScriptInsertionPoint({
+      document,
+      originalSource,
+    });
+
+    // create Paragon script to inject into the HTML document
+    const paragonScript = `
+      <script type="text/javascript">
+        var PARAGON_THEME = ${JSON.stringify(scriptContents, null, 2)};
+      </script>
+    `;
+
+    // insert the Paragon script into the HTML document
+    const newSource = new sources.ReplaceSource(
+      new sources.RawSource(originalSource),
+      'index.html',
+    );
+    newSource.insert(scriptInsertionPoint, this.minifyScript(paragonScript));
+    return newSource;
+  }
+
+  apply(compiler) {
+    compiler.hooks.thisCompilation.tap('ParagonWebpackPlugin', (compilation) => {
+      compilation.hooks.processAssets.tap(
+        {
+          name: 'ParagonWebpackPlugin',
+          stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONS,
+          additionalAssets: true,
+        },
+        () => {
+          const file = compilation.getAsset('index.html');
+          if (!file) {
+            return;
+          }
+          const {
+            coreCssAsset: paragonCoreCssAsset,
+            themeVariantCssAssets: paragonThemeVariantCssAssets,
+          } = this.getParagonCssAssetsFromCompilation(compilation);
+
+          const {
+            coreCssAsset: brandCoreCssAsset,
+            themeVariantCssAssets: brandThemeVariantCssAssets,
+          } = this.getParagonCssAssetsFromCompilation(compilation, { isBrandOverride: true });
+
+          let scriptContents;
```

Comment:
Do you think it would be simpler to just define `const scriptContents = {}` and remove `createEmptyScriptContentsIfUndefined` function?

Reply:
D'oh, it would seem that way! Just a tad bit over-engineered it seems 😜